Add endpoint that exposes realm connection state and upload progress

fealebenpae opened the pull request.

fealebenpae committed:
Add endpoint that exposes realm connection state and upload progress

nirinchev commented on src/service.ts:
```diff
@@ -1079,6 +1097,16 @@ export class GraphQLService {
 
+    this.state[path] = <any>{ connection: Realm.Sync.ConnectionState.Connected };
+
+    realm.syncSession.addConnectionNotification((state) => {
```
Should we remove these listeners when the Realm is closed?

fealebenpae replied:
They're on the session, and that outlives the realm. Besides, when the session is destroyed it lets go of the listeners anyway.

nirinchev replied:
My concern is that this will add a listener on every request. Perhaps node.js is clever and will recognize that this listener has already been added to the session, but if that's the case, it's not immediately obvious to me. If it's not clever though, I can easily see a long-lived Realm keeping a collection of thousands of listeners doing the same thing.

nirinchev committed:
Some fixes:
- Rename state to avoid clashing with IService.state
- Change route to be connectionStatus rather than synchronizing
- Move getConnection status above the graphql route to make sure it's handled properly
- Don't return 404s

ianpward commented:
@fealebenpae @nirinchev Do we think we can complete this for this coming week? I have a meeting with Customer on Wednesday

fealebenpae commented:
@nirinchev and I had a brain-storming session today on how to proceed with the leaking listeners issue and we'll try to get this reviewed and merged by tomorrow so we can cut a release for Customer.

fealebenpae committed:
remove connection listeners when closing realms

nirinchev commented on src/service.ts:
```diff
@@ -1079,6 +1123,26 @@ export class GraphQLService {
       user,
     });
 
+    this.connectionStates[path] = {
```
We should not override this if it exists.

fealebenpae committed:
do not clobber the connection state when opening a realm

ianpward commented:
can we merge this to master please? its been sitting since nov29 cc. @geragray

nirinchev commented:
It has some issues, which is why it hasn't been merged.
